wasm.md

```diff
@@ -123,12 +123,30 @@ The sorts `EmptyStmt` and `EmptyStmts` are administrative so that the empty list
 
 ### Traps
 
-When a single value ends up on the instruction stack (the `<k>` cell), it is moved over to the value stack (the `<valstack>` cell).
-If the value is the special `undefined`, then `trap` is generated instead.
+`trap` is the error mechanism of Wasm.
+Traps cause all execution to halt, and can not be caught from within Wasm.
+We emulate this by consuming everything in the `<k>` cell that is not a `Stmt`.
+Statements are not part of Wasm semantics, but rather of the embedder, and is where traps can be caught.
+Thus, a `trap` "bubbles up" (more correctly, to "consumes the continuation") until it reaches a statement which is not an `Instr` or `Def`.
 
 ```k
     syntax Instr ::= "trap"
  // -----------------------
+    rule <k> trap ~> (L:Label      => .) ... </k>
+    rule <k> trap ~> (F:Frame      => .) ... </k>
+    rule <k> trap ~> (I:Instr      => .) ... </k>
+    rule <k> trap ~> (IS:Instrs    => .) ... </k>
+    rule <k> trap ~> (D:Defn       => .) ... </k>
+    rule <k> trap ~> (DS:Defns     => .) ... </k>
+    rule <k> trap ~> ((module  _)  => .) ... </k>
+
+    rule <k> trap ~> (S:Stmt SS:Stmts => S ~> SS) ... </k>
```

Comment:
actually it is a little twisty to me.
So if a `trap` is followed by `(invoke 0)`, it will be skipped because `(invoke 0)` is an `Instr`. But if a `trap` is followed by `(invoke "abc")`, it will be stopped because `(invoke "abc")` is an `Auxil`. Is it really what we want?

Reply:
Isn't that because `invoke 0` is part of Wasm semantics (it's an administrative instruction), whereas `invoke "abc"` is an `Auxil`, outside of Wasm semantics.

If we want, we can add a rule in `TEST` that says that traps eat `Auxil`, too, as long as they are not assertions, but I think that's wrong. If a `trap` is followed by anything but an `#assertTrap`, then execution will halt, right?

The point is: The spec says that traps propagate to the top in Wasm, and can't be caught from within it. These rules make sure that happens. What happens outside of Wasm (once we leave a module instantiation) is up to us.

Reply:
I was just a little unsure about it. It should be fine!